src/string.rs

```diff
@@ -0,0 +1,129 @@
+use std::{
+    fmt::Display,
+    ops::{Deref, DerefMut},
+    str::{from_utf8_unchecked, from_utf8_unchecked_mut},
+};
+
+#[derive(Debug)]
+pub enum CowStr<'s> {
+    Owned(String),
+    Borrowed(&'s str),
+    Inlined(InlineStr),
+}
+
+impl<'s> Deref for CowStr<'s> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Self::Owned(ref s) => s,
+            Self::Borrowed(s) => s,
+            Self::Inlined(s) => s.deref(),
+        }
+    }
+}
+
+impl<'s> AsRef<str> for CowStr<'s> {
+    fn as_ref(&self) -> &str {
+        self.deref()
+    }
+}
+
+impl<'s> From<&'s str> for CowStr<'s> {
+    fn from(value: &'s str) -> Self {
+        CowStr::Borrowed(value)
+    }
+}
+
+impl<'s> From<String> for CowStr<'s> {
+    fn from(value: String) -> Self {
+        CowStr::Owned(value)
+    }
+}
+
+impl<'s> Clone for CowStr<'s> {
+    fn clone(&self) -> Self {
+        match self {
+            CowStr::Owned(s) => match InlineStr::try_from(&**s) {
+                Ok(inline) => CowStr::Inlined(inline),
+                Err(_) => CowStr::Owned(s.clone()),
+            },
+            CowStr::Borrowed(s) => CowStr::Borrowed(s),
+            CowStr::Inlined(s) => CowStr::Inlined(*s),
+        }
+    }
+}
+
+impl<'s> PartialEq for CowStr<'s> {
+    fn eq(&self, other: &Self) -> bool {
+        self.deref() == other.deref()
+    }
+}
+
+impl<'s> Eq for CowStr<'s> {}
+
+impl<'s> Display for CowStr<'s> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.deref())
+    }
+}
+
+impl<'s, 'a> FromIterator<&'a str> for CowStr<'s> {
+    fn from_iter<T: IntoIterator<Item = &'a str>>(iter: T) -> Self {
+        CowStr::Owned(FromIterator::from_iter(iter))
+    }
+}
+
+const MAX_INLINE_STR_LEN: usize = 3 * std::mem::size_of::<isize>() - 2;
+
+pub struct Error;
+
+#[derive(Clone, Copy, Debug)]
+pub struct InlineStr {
```

Comment:
do we need to have a dedicated type? Can't it be handled internally within the CowStr? I.e. CowStr::Inline is used as long as it fits, otherwise CowStr::Owned is used.

Reply:
I did it originally to allow a bit of ergonomics, like the `TryFrom`, etc, and to make the API a bit friendlier. However, seeing as we want this to be as `String`-like as possible, users shouldn't really need to interact with the inner values at all, so storing `InlineStr` internally saves some lines of code. Changed now!